### DISCUSSION TITLE
Index on foreign keys in postgres

### DISCUSSION THREAD

**Discussion by simonstaffgenius** [prisma/prisma#25783] [Q&A] [answered]
### Question

When creating relations between tables a foreign key will be created when using the default `relationMode = "foreignKeys"`.
But neither postgres nor prisma will create an **index** for these referential columns.

```
model User {
  id    Int    @id @default(autoincrement())
  posts Post[]
}

model Post {
  id       Int  @id @default(autoincrement())
  author   User @relation(fields: [authorId], references: [id])
  authorId Int // I would expect an index to be created here.

 // @@index([authorId]) // For now you have to create those indices manually
}
```

For join queries this could potentially speed up the query especially when joining large tables.

I found this in the docs:

> In relational databases that use foreign key constraints, the database usually also implicitly creates an index for the foreign key columns. For example, [MySQL will create an index on all foreign key columns](https://dev.mysql.com/doc/refman/8.0/en/constraint-foreign-key.html#:~:text=MySQL%20requires%20that%20foreign%20key%20columns%20be%20indexed%3B%20if%20you%20create%20a%20table%20with%20a%20foreign%20key%20constraint%20but%20no%20index%20on%20a%20given%20column%2C%20an%20index%20is%20created.). This is to allow foreign key checks to run fast and not require a table scan.

Unfortunately Postgres will not create an index implicitly like MySQL does.
Wouldn't it make sense to automatically create those indices on foreign columns directly through prisma if the underlying database is not handling it?

**Comment by ludralph** [comment]
Hi @simonstaffgenius 👋 

Thank you for raising this. This topic has previously been addressed in this open [issue](https://github.com/prisma/prisma/issues/10611). We introduced an [index validation](https://www.prisma.io/docs/orm/prisma-schema/data-model/relations/relation-mode#index-validation) warning in Prisma ORM versions 4.7.0 and later that alerts users when a relation doesn't have an index on the relation scalar field  If you're using the Prisma VS Code Extension. For now, the recommended approach is to manually add indexes to your foreign key columns, as you've shown in your example.

**Comment by simonstaffgenius** [comment]
Hi @ludralph 
Thank you very much for the quick response.

I saw this issue and the discussion around it, but was under the impression that the warning will only be raised if you use `relationMode = prisma`?

So as far as I understand your answer, no matter which `relationMode` you're using, creating indexes on relation scalar fields have to be managed manually for now.

**Reply by ludralph** [reply, marked as answer]
Yes, you're correct in your understanding. Creating indexes on relation scalar fields has to be managed manually for now. The automatic warnings are indeed specific to `relationMode = "prisma"`, as this mode doesn't use foreign keys at all. For PostgreSQL users using the default foreignKeys mode, it's a good practice to manually add indexes to foreign key columns for performance reasons, even though Prisma won't warn you about their absence.

**Reply by sricks** [reply]
Is there any way to get warnings about this when `relationMode="foreignKeys"` with postgres?

There are definitely valid reasons not to create an index in this situation, but IME they are rare. IMO those should require an explicit lint disable comment.
